types/react/index.d.ts

```diff
@@ -90,6 +90,32 @@ declare namespace React {
     type RefCallback<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"];
     type Ref<T> = RefCallback<T> | RefObject<T> | null;
     type LegacyRef<T> = string | Ref<T>;
+    /**
+     * Gets the instance type for a React element. The instance will be different for various component types:
+     *
+     * - React class components will be the class instance. So if you had `class Foo extends React.Component<{}> {}`
+     *   and used `React.ElementRef<typeof Foo>` then the type would be the instance of `Foo`.
+     * - React stateless functional components do not have a backing instance and so `React.ElementRef<typeof Bar>`
+     *   (when `Bar` is `function Bar() {}`) will give you the `undefined` type.
+     * - JSX intrinsics like `div` will give you their DOM instance. For `React.ElementRef<'div'>` that would be
+     *   `HTMLDivElement`. For `React.ElementRef<'input'>` that would be `HTMLInputElement`.
+     *
+     * `C` must be the type _of_ a React component so you need to use typeof as in React.ElementRef<typeof MyComponent>.
+     */
+    type ElementRef<
+        C extends
+            | ComponentClass<any>
+            | FunctionComponent<any>
+            | keyof JSX.IntrinsicElements
```

Comment:
How about for readability:
```
type ElementCType = ComponentClass<any> | FunctionComponent<any> | keyof JSX.IntrinsicElements;
type ElementRef<C extends ElementCType> = ...
```

Reply:
I didn’t want to add intermediate types into this namespace, but if people agree then I’m definitely not against it.

Reply:
This already exists `JSXElementConstructor`.

Reply:
I tried it, but didn’t feel like readability improved by using a type alias in this first part but then use different types below. I think if there were very narrowly defined type aliases for these then that might be helpful, but that would introduce types that don’t exist upstream and so I’d like an emphatic “yes please” before I do that.